src/base/scoped_sched_boost.cc

```diff
@@ -0,0 +1,231 @@
+  // It is possible that we previously added the wrongly configured policy,
+  // that wasn't the max policy. In that case the policy will be validated now,
+  // in 'RecalcAndUpdatePrio'.
+  // This for loop makes us ignore all wrongly configured policies and fallback
+  // to the first correct (or the initial one).
+  for (;;) {
+    if (auto res = RecalcAndUpdatePrio(); !res.ok()) {
+      PERFETTO_ELOG("%s", res.c_message());
+    } else {
+      break;
+    }
+  }
+}
+
+Status ThreadMgr::RecalcAndUpdatePrio() {
+  if (prios_.empty()) {
+    return os_manager_->SetSchedConfig(initial_config_);
+  }
+  // TODO(ktimofeev): Check previously set prio to skip unnecessary syscall?
+  auto max_prio = std::max_element(prios_.begin(), prios_.end());
+  SchedOsManager::SchedOsConfig os_config{};
+  switch (max_prio->policy) {
+    case SchedPolicyAndPrio::Policy::kSchedOther:
+      os_config = SchedOsManager::SchedOsConfig{
+          SCHED_OTHER, 0, -1 * static_cast<int>(max_prio->prio)};
+      break;
+    case SchedPolicyAndPrio::Policy::kSchedFifo:
+      os_config = SchedOsManager::SchedOsConfig{
+          SCHED_FIFO, static_cast<int>(max_prio->prio), 0};
+      break;
+  }
+  Status res = os_manager_->SetSchedConfig(os_config);
+  if (!res.ok()) {
+    prios_.erase(max_prio);
+    return res;
+  }
+  return OkStatus();
+}
+
+void ThreadMgr::ResetForTesting(SchedOsManager* os_manager) {
+  os_manager_ = os_manager;
+  initial_config_ = os_manager->GetCurrentSchedConfig().value();
+  prios_.clear();
+}
+
+}  // namespace
+
+#if PERFETTO_BUILDFLAG(PERFETTO_OS_LINUX) || \
+    PERFETTO_BUILDFLAG(PERFETTO_OS_ANDROID)
+
+SchedOsManager* SchedOsManager::GetInstance() {
+  static auto* instance = new SchedOsManager();
+  return instance;
+}
+
+Status SchedOsManager::SetSchedConfig(const SchedOsConfig& arg) {
+  sched_param param{};
+  param.sched_priority = arg.rt_prio;
+  int ret = sched_setscheduler(kCurrentPid, arg.policy, &param);
+  if (ret == -1) {
+    return ErrStatus("sched_setscheduler(%d, %d) failed (errno: %d, %s)",
+                     arg.policy, arg.rt_prio, errno, strerror(errno));
+  }
+  if (arg.rt_prio == 0) {
+    ret = setpriority(PRIO_PROCESS, kCurrentPid, arg.nice);
+    if (ret == -1) {
+      return ErrStatus("setpriority(%d) failed (errno: %d, %s)", arg.nice,
+                       errno, strerror(errno));
+    }
+  }
+  return OkStatus();
+}
+
+StatusOr<SchedOsManager::SchedOsConfig> SchedOsManager::GetCurrentSchedConfig()
+    const {
+  int policy = sched_getscheduler(kCurrentPid);
+  if (policy == -1) {
+    return ErrStatus("sched_getscheduler failed (errno: %d, %s)", errno,
+                     strerror(errno));
+  }
+  sched_param param{};
+  if (sched_getparam(kCurrentPid, &param) == -1) {
+    return ErrStatus("sched_getparam failed (errno: %d, %s)", errno,
+                     strerror(errno));
+  }
+  int nice = 0;
+  if (param.sched_priority == 0) {
```

Comment:
add a comment as this is a bit subtle. explain that SCHED_OTHER/BATCH make use of the niceness vaue